src/number_theory/modular_forms/modular_forms.lean

```diff
@@ -0,0 +1,391 @@
+/-
+Copyright (c) 2022 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck, David Loeffler
+-/
+import algebra.module.submodule.basic
+import analysis.complex.upper_half_plane
+import linear_algebra.general_linear_group
+import linear_algebra.special_linear_group
+import algebra.direct_sum.ring
+import number_theory.modular
+import geometry.manifold.mfderiv
+
+/-!
+# Modular forms
+
+This file defines modular forms and proves some basic properties about them.
+
+We begin by defining the `slash_k` operator on the space of functions `ℍ → ℂ` and from this
+define the notion of weakly modular form.
+
+We then define `bounded_at_infinity` and `zero_at_infinity`. Finally we construct the vector
+space of modular forms and prove that the product of two modular forms is a modular form
+(of higher weight).
+-/
+
+universes u v
+
+open complex
+
+open_locale topological_space manifold upper_half_plane
+
+noncomputable theory
+
+local notation `ℍ'` := (⟨upper_half_plane.upper_half_space ,
+ upper_half_plane.upper_half_plane_is_open⟩: topological_space.opens ℂ)
+
+local prefix `↑ₘ`:1024 := @coe _ (matrix (fin 2) (fin 2) _) _
+
+local notation `GL(` n `, ` R `)`⁺ := matrix.GL_pos (fin n) R
+
+local notation `SL(` n `, ` R `)` := matrix.special_linear_group (fin n) R
+
+variable (M : GL(2, ℝ)⁺)
+
+/--The weight `k` action of `GL(2, ℝ)⁺` on functions `f : ℍ → ℂ`. -/
+def slash_k : ℤ → GL(2, ℝ)⁺ → (ℍ → ℂ) → (ℍ → ℂ) := λ k γ f,
```

Comment:
I think is more mathlibby to call it `slash`, since `k` is a variable.

In any case I am wondering whether it is convenient to bundle it somehow.

Reply:
I've changed the name, but I'm not sure how to bundle it?

Reply:
Ok I'm trying to figure out how to do this. I proved:

```
def slash_action (k : ℤ) : mul_action (mul_opposite GL(2, ℝ)⁺) (ℍ → ℂ) :=
{ smul := λ A, slash k (mul_opposite.unop A),
  one_smul := by {dsimp, apply slash_mul_one,},
  mul_smul := by { intros A B f, dsimp,  simp  [slash_right_action],},}
```

But I don't know how to use a "parametrized" instance. I wanted to define some notation for this, like : 

`localized "notation f ` ∣[`:100 k `]`:0 γ :100 := (slash_action k).smul γ f" in modular_form` but this doesn't work. Any ideas on how to do this?

Reply:
I don't know, but I think @alexjbest is a reasonable person to ask.

Reply:
Also, the only way I can see to make right actions is using the `mul_opposite`, so there will be some extra coercions needed to make put elements into the opposite group. But hopefully this isn't too annoying ? I'm not sure if there is a better way to do this.

Reply:
So I rememebered asking about this here: https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/notation.20for.20a.20family.20of.20mul_actions 

And I've now tried the `with_k` option, but as was mentioned there, the `with_k` becomes a pain, as does the having `mul_opposites` around. I could try making a new class for this, but I don't know if that would help

Reply:
OK so I've now tried making a new class for `slash_actions` which contains a bunch of the useful properties. It still needs some work, and if we are happy with it, it will probably go in a separate file/PR. But is this more like what you had in mind?

Reply:
One thing that might be useful about this, is that it might help sort out all the `coe` issues comming from swapping between congruence subgroups, SL2(Z) and GL_pos. It should also bhopefully by easy to extend it to add characters for modular forms later on.

Reply:
I don't have any strong opinion, but if you have the impression that making a new class works well just go for it. At the end we are just looking for an approach that makes things as painless as possible.

Reply:
There is nothing magic about `charted_space`. Someone has provided [topological_space.opens.charted_space](https://leanprover-community.github.io/mathlib_docs/geometry/manifold/charted_space.html#topological_space.opens.charted_space) and that's why Lean know that an open of a `charted_space` is a `charted_space`.

Reply:
Yeah I wanted to do the same here with the `induced_slash_action`. I tried making it an instance but it then wouldn't work when I had the monoid hom. So I've instead made it a defn. In any case, I think this works well enough now. Should I split this `slash_action` into a new PR?

Reply:
Pre-empting the answer yes, I've made #15007 with slash_actions

Reply:
I am a little busy these days, but I will try to have a look as soon as possible.